src/librustc_jobserver/lib.rs

```diff
@@ -0,0 +1,292 @@
+    }
+
+    fn acquire_token_from_request(&self) {
+        // We essentially duplicate acquire_token below but never emit a request
+        // for additional tokens.
+        log::trace!("{:?} acquiring token from request", std::thread::current().id());
+        let mut tokens = self.tokens.lock().unwrap();
+        loop {
+            if tokens.used == 0 {
+                log::trace!("{:?} acquired implicit token", std::thread::current().id());
+                tokens.used = 1;
+                return;
+            } else if tokens.pending > 0 {
+                // forget the token so as to not drop
+                tokens.pending -= 1;
+                tokens.used += 1;
+                log::trace!("{:?} acquired real token", std::thread::current().id());
+                return;
+            }
+            tokens = self.waiter.wait(tokens).unwrap();
+        }
+    }
+
+    fn acquire_token(&self) {
+        log::trace!("{:?} acquiring token", std::thread::current().id());
+        let mut tokens = self.tokens.lock().unwrap();
+        let mut has_requested = false;
+        loop {
+            if tokens.used == 0 {
+                log::trace!("{:?} acquired implicit token", std::thread::current().id());
+                tokens.used = 1;
+                return;
+            } else if tokens.pending > 0 {
+                // forget the token so as to not drop
+                tokens.pending -= 1;
+                tokens.used += 1;
+                log::trace!("{:?} acquired real token", std::thread::current().id());
+                return;
+            }
+            if !has_requested {
+                has_requested = true;
+                self.request_token();
+            }
+            tokens = self.waiter.wait(tokens).unwrap();
+        }
+    }
+
+    fn release_token(&self) {
+        log::trace!("{:?} releasing token", std::thread::current().id());
+        let mut tokens = self.tokens.lock().unwrap();
+        tokens.used -= 1;
+        if tokens.used == 0 {
+            log::trace!("{:?} released implicit token", std::thread::current().id());
+            self.waiter.notify_one();
+        } else {
+            log::trace!("{:?} released real token", std::thread::current().id());
+            if should_notify() {
+                eprintln!(
+                    "{}",
+                    as_json(&JobserverNotification { jobserver_event: Event::Release })
+                );
+            } else {
+                GLOBAL_CLIENT.release_raw().unwrap();
+            }
+        }
+    }
+
+    pub fn request_token(&self) {
+        log::trace!("{:?} requesting token", std::thread::current().id());
+        // Just notify, don't actually acquire here.
+        notify_acquiring_token();
+        self.helper.lock().unwrap().request_token();
+    }
+}
+
+#[must_use]
+pub struct Acquired(());
+
+impl Drop for Acquired {
+    fn drop(&mut self) {
+        release_thread();
+    }
+}
+
+pub fn helper_channel() -> &'static Helper {
```

Comment:
If possible could this be hidden? I'd like to ideally shrink the public API surface area of this module as much as possible so there's as little to audit as we need.

Reply:
This is just exposing a `request_token()` function, and needs to tie it to some struct so I think we can't hide it (or at least I wasn't able to come up with an obvious way).